### PR TITLE
[codegen] Add max(half, half) support when enable fp16

### PR DESCRIPTION
**Pull request by ZQPei**
Fix the following error when compiled with float16 model in cuda.
```
/tmp/tmpz_0pydlm/my_kernel.cu(9890): error: more than one instance of overloaded function "max" matches the argument list:
            function "max(int, int)"
            function "max(unsigned int, unsigned int)"
            function "max(int, unsigned int)"
            function "max(unsigned int, int)"
            function "max(long, long)"
            function "max(unsigned long, unsigned long)"
            function "max(long, unsigned long)"
            function "max(unsigned long, long)"
            function "max(long long, long long)"
            function "max(unsigned long long, unsigned long long)"
            function "max(long long, unsigned long long)"
            function "max(unsigned long long, long long)"
            function "max(float, float)"
            argument types are: (half, __half)
```
Please check!

### CODE REVIEW

**Comment by cchung100m** [comment]
Hi @ZQPei 

Please check the CI error

> docker/bash.sh tvmai/ci-lint:v0.51 ./tests/scripts/task_lint.sh

```
Makefile:70: recipe for target 'cpplint' failed

make: *** [cpplint] Error 1

script returned exit code 2
```

**Comment by ZQPei** [comment]
Hi, @cchung100m 
I do not understand what the `CI error` might be.
The error is caused by `cpplint failed` .
What is cpplint? Can you give me a hint?
I will do what I can to fix it.

BTW,
This commit works well on all of my machines, and I do believe it can work on the others.
There should not be any error when compiling TVM.

**Comment by ZQPei** [comment]
@cchung100m 
I tried run `sh tests/scripts/task_lint.sh`
and it shows the following
```
Check file types...
check_file_type.py: all checks passed..
Check ASF license header...
Error: Unable to access jarfile /bin/apache-rat.jar
```
What does this mean?

**Comment by ZQPei** [comment]
Hi @cchung100m 
Now I have passed all 5 checks
please check
Thanks

**Comment by cchung100m** [comment]
Hi @ZQPei 

For your PR, besides `codegen_cuda.cc`, we should also need to add the unittest in `tests/python/unittest/test_codegen_cuda.py`.

**Comment by ZQPei** [comment]
How to add a unittest？
Do I need to do that？

**Comment by cchung100m** [comment]
Hi @ZQPei 

In my opinion, please refer to the [post](https://discuss.tvm.ai/t/failed-to-convert-mxnet-float16-model-to-tvm/1557/2) at the tvm forum to develop your testing script.

hope helps,

**Comment by ZQPei** [comment]
@cchung100m 
I have wrote a unit test for this PR.
Test was passed on my machine.
Should I add this code to TVM?

Here is the unit test.
```python
def test_cuda_vector_max():
    num_thread = 8
    target = 'cuda'
    def check_vector_max(ctx, n, dtype):
        if not tvm.gpu(0).exist or not tvm.module.enabled("cuda"):
            print("skip because cuda is not enabled..")
            return
        if dtype == "float16" and not have_fp16(tvm.gpu(0).compute_version):
            print("skip because gpu does not support fp16")
            return
        A = tvm.placeholder((n,), name='A', dtype=dtype)
        B = tvm.placeholder((n,), name='B', dtype=dtype)
        C = tvm.compute((n,), lambda i: tvm.max(A[i], B[i]), name='C')
        s = tvm.create_schedule(C.op)
        bx, tx = s[C].split(C.op.axis[0], factor=num_thread)
        s[C].bind(bx, tvm.thread_axis("blockIdx.x"))
        s[C].bind(tx, tvm.thread_axis("threadIdx.x"))
        fun = tvm.build(s, [A,B,C], "cuda", name="vector_max")

        np_a = np.random.uniform(size=n).astype(dtype)
        np_b = np.random.uniform(size=n).astype(dtype)
        np_c = np.maximum(np_a, np_b)
        a = tvm.nd.empty((n,), A.dtype, ctx).copyfrom(np_a)
        b = tvm.nd.empty((n,), B.dtype, ctx).copyfrom(np_b)
        c = tvm.nd.empty((n,), C.dtype, ctx)
        np.testing.assert_equal(c.asnumpy(), np_c)

    ctx = tvm.context(target, 0)
    check_vector_max(ctx, 10, "float32")
    check_vector_max(ctx, 10, "float16")
```

File: src/codegen/codegen_cuda.cc [hunk 1]
```diff
@@ -50,6 +50,8 @@ void CodeGenCUDA::AddFunction(LoweredFunc f) {
 std::string CodeGenCUDA::Finish() {
   if (enable_fp16_) {
     decl_stream << "#include <cuda_fp16.h>\n";
+    decl_stream << "__device__ half max(const half a, const half b)\n"
```
Comment (anijain2305) [inline src/codegen/codegen_cuda.cc]: Do we know which operators we have to overload as such? "max" is one of them. Do we need others?

Comment (ZQPei) [reply src/codegen/codegen_cuda.cc]: For now, I only find `max` that need to be overloaded.

BTW, I have a question about the checks.
Why this commit cannot be built today? It was successful yesterday.

**Comment by ZQPei** [comment]
Hi @cchung100m @anijain2305 
The commit to `codegen_cuda.cc` is passed yesterday.
However, the commit of adding unittest to `test_codegen_cuda.py` was failed today.
Can you help me to find out what caused the check failed?

**Comment by vinx13** [comment]
@ZQPei please also add the test case to this PR

File: src/codegen/codegen_cuda.cc (hunk 1 above)
Comment (anijain2305) [reply src/codegen/codegen_cuda.cc]: Hi, we saw more failures while trying to run full resnet

https://github.com/dmlc/tvm/issues/3816#issuecomment-524067090

I think, we are missing all reduce ops. Will it be possible for you to help with this? (In a separate PR, this one is good to go)

**Comment by cchung100m** [comment]
Hi @ZQPei 

Please first post in [discuss.tvm.ai](https://discuss.tvm.ai) and provide more details of what you're doing. Currently, it is not clear where the problem is.

**Comment by anijain2305** [comment]
HI, just a ping to get this in :)

**Comment by tqchen** [comment]
superceded by https://github.com/dmlc/tvm/pull/4056. Thanks @ZQPei  for your contribution

**Pull request closed without merging**